File: bot.js
const { spawn, exec } = require('child_process');
const TelegramBot = require('node-telegram-bot-api');
const got = require('got');
const fs = require('fs');

const CONFIG_PATH = 'config.json';

if (!fs.existsSync(CONFIG_PATH)) {
    const defaultConfig = {
        TELEGRAM_TOKEN: "TOKEN_TELEGRAM",
        CHAT_ID: "CHAT_ID",
        IPDB_API_KEY: "НА_ЭТО_ХУЙ_ЗАБЕЙ",
        SSH_SERVICE: "ssh.service"  // не советую менять
    };

    fs.writeFileSync(CONFIG_PATH, JSON.stringify(defaultConfig, null, 4), 'utf8');
    console.log("Конфигурационный файл 'config.json' был создан с настройками по умолчанию.");
    process.exit()
}

const config = JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf8'));

const bot = new TelegramBot(config.TELEGRAM_TOKEN, { polling: false });

function sendTelegramMessage(message) {
    bot.sendMessage(config.CHAT_ID, message).catch(err => console.error('Ошибка при отправке сообщения:', err));
}

function blockIP(ip, reason) {
    const command = `sudo iptables -I INPUT -s ${ip} -j DROP`;
    exec(command, (error) => {
        if (error) {
            console.error(`Ошибка блокировки IP ${ip}:`, error);
        } else {
            sendTelegramMessage(`IP ${ip} заблокирован на 30 дней.\nПричина: ${reason}`);
        }
    });
}

async function getIPInfo(ip) {
    try {
        const response = await got(`http://ip-api.com/json/${ip}`).json();
        const { country, city, as, regionName } = response;
        return {
            isChinese: country === 'China',
            country,
            city,
            as,
            region: regionName
        };
    } catch (error) {
        console.error('Ошибка при проверке IP через ip-api:', error);
        return null;
    }
}

const journal = spawn('journalctl', ['-u', config.SSH_SERVICE, '-f', '-n', '0']);

journal.stdout.on('data', async (data) => {
    const output = data.toString();
    const match = output.match(/Failed password for (?:invalid user )?(\w+) from ([\d.]+) port \d+ ssh2/);

    if (match) {
        const username = match[1];
        const ipAddress = match[2];

        const ipInfo = await getIPInfo(ipAddress);

        if (ipInfo) {
            const { isChinese, country, city, as, region } = ipInfo;

            let message = `Попытка входа с IP: ${ipAddress}\nUsername: ${username}\n`;
            message += `Страна: ${country}\nГород: ${city}\nAS: ${as}\nРегион: ${region}\n`;

            if (isChinese) {
                blockIP(ipAddress, 'Китайский IP');
                message += `⚠️ Этот IP был заблокирован, так как он из Китая.\nIP заблокирован на 30 дней.`;
            }

            sendTelegramMessage(message);
        }
    }
});

journal.stderr.on('data', (data) => {
    console.error(`Ошибка: ${data}`);
});

journal.on('close', (code) => {
    console.log(`Процесс journalctl завершился с кодом ${code}`);
});

File: bot.py
import json
import re
import subprocess
import asyncio
import aiohttp
from telegram import Bot

with open('config.json', 'r') as f:
    config = json.load(f)

bot = Bot(token=config['TELEGRAM_TOKEN'])

async def send_telegram_message(message):
    try:
        await bot.send_message(chat_id=config['CHAT_ID'], text=message)
    except Exception as e:
        print(f"Ошибка при отправке сообщения в Telegram: {e}")

def block_ip(ip):
    command = f"sudo iptables -I INPUT -s {ip} -j DROP"
    try:
        subprocess.run(command, shell=True, check=True)
        print(f"IP {ip} заблокирован.")
    except subprocess.CalledProcessError as e:
        print(f"Ошибка при блокировке IP {ip}: {e}")

async def get_ip_info(ip):
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(f"http://ip-api.com/json/{ip}") as response:
                data = await response.json()
                country = data.get("country")
                city = data.get("city")
                as_name = data.get("as")
                region = data.get("regionName")
                return country, city, as_name, region
    except Exception as e:
        print(f"Ошибка при получении информации о IP {ip}: {e}")
        return None, None, None, None

def monitor_ssh_login():
    journal = subprocess.Popen(['journalctl', '-u', 'ssh.service', '-f', '-n', '0'], stdout=subprocess.PIPE)

    for output in iter(journal.stdout.readline, b''):
        output = output.decode('utf-8')
        match = re.search(r'Failed password for (?:invalid user )?(\w+) from ([\d.]+) port \d+ ssh2', output)

        if match:
            username = match.group(1)
            ip_address = match.group(2)

            country, city, as_name, region = asyncio.run(get_ip_info(ip_address))

            message = f"Попытка входа с IP: {ip_address}\nUsername: {username}\n"
            message += f"Страна: {country}\nГород: {city}\nAS: {as_name}\nРегион: {region}\n"

            asyncio.run(send_telegram_message(message))

            if country == "China":
                block_ip(ip_address)
                message = f"⚠️ Этот IP был заблокирован, так как он из Китая.\nIP заблокирован на 30 дней."
                asyncio.run(send_telegram_message(message))

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_in_executor(None, monitor_ssh_login)
    loop.run_forever()
